src/System.Text.Primitives/System/Text/Encoding/Utf8/Utf8Encoder.cs

```diff
@@ -51,11 +51,44 @@ internal static class Utf8Encoder
         /// <param name="bytesConsumed">On exit, contains the number of code points that were consumed from the UTF-16 character span.</param>
         /// <param name="charactersWritten">An output parameter to store the number of characters written to <paramref name="utf16"/></param>
         /// <returns>True if the input buffer was fully encoded into the output buffer, otherwise false.</returns>
-        public static bool TryDecode(ReadOnlySpan<byte> utf8, Span<char> utf16, out int bytesConsumed, out int charactersWritten)
+        public static unsafe bool TryDecode(ReadOnlySpan<byte> utf8, Span<char> utf16, out int bytesConsumed, out int charactersWritten)
         {
             bytesConsumed = 0;
             charactersWritten = 0;
 
+            if (Vector.IsHardwareAccelerated)
+            {
+                fixed (byte* outerUtf8Ptr = &utf8.DangerousGetPinnableReference())
```

Comment:
You can avoid unsafe and fixed blocks by using by-ref and the Unsafe static helper class (which you are already using anyways). The assembly code that gets generated ends up being almost identical to the pointer version without the setup cost of the fixed block.

It is the pattern we are trying to push now since it leads to slightly safer and neater code (IMO).